pkg/icingadb/history/sync.go

```diff
@@ -354,32 +380,210 @@ func userNotificationStage(ctx context.Context, s Sync, key string, in <-chan re
+
+		makeEntity := func(values map[string]interface{}) (database.Entity, error) {
+			val, err := structifier(values)
+			if err != nil {
+				return nil, errors.Wrapf(err, "can't structify values %#v for %q", values, key)
+			}
+
+			entity, ok := val.(database.Entity)
+			if !ok {
+				return nil, fmt.Errorf("structifier returned %T which does not implement database.Entity", val)
+			}
+
+			return entity, nil
+		}
+
+		backlogLastId := ""
+		backlogMsgCounter := 0
+
+		const backlogTimerMinInterval, backlogTimerMaxInterval = 10 * time.Millisecond, time.Minute
+		backlogTimerInterval := backlogTimerMinInterval
+		backlogTimer := time.NewTimer(backlogTimerInterval)
+		_ = backlogTimer.Stop()
+
+		for {
+			select {
+			case msg, ok := <-in:
+				if !ok {
+					return nil
+				}
+
+				// Only submit the entity directly if there is no backlog.
+				// The second check covers a potential corner case if the XRANGE below races this stream.
+				if backlogLastId != "" && backlogLastId != msg.ID {
+					continue
+				}
+
+				entity, err := makeEntity(msg.Values)
+				if err != nil {
+					return err
+				}
+
+				if callback(entity) {
+					out <- msg
+					telemetry.Stats.Callback.Add(1)
+					backlogLastId = ""
+				} else {
+					backlogLastId = msg.ID
+					backlogMsgCounter = 0
+					backlogTimerInterval = backlogTimerMinInterval
+					_ = backlogTimer.Reset(backlogTimerInterval)
+					s.logger.Warnw("Failed to submit entity to callback, entering into backlog",
+						zap.String("key", key),
+						zap.String("id", backlogLastId))
+				}
+
+			case <-backlogTimer.C:
+				if backlogLastId == "" { // Should never happen.
+					return fmt.Errorf("backlog timer logic for %q was called while backlogLastId was empty", key)
+				}
+
+				logger := s.logger.With(
+					zap.String("key", key),
+					zap.String("last-id", backlogLastId))
+
+				logger.Debug("Trying to advance backlog of callback elements")
+
+				xrangeCmd := s.redis.XRangeN(ctx, "icinga:history:stream:"+key, backlogLastId, "+", 2)
+				msgs, err := xrangeCmd.Result()
+				if err != nil {
+					return errors.Wrapf(err, "XRANGE %q to %q on stream %q failed", backlogLastId, "+", key)
+				}
+
+				if len(msgs) < 1 || len(msgs) > 2 {
+					return fmt.Errorf("XRANGE %q to %q on stream %q returned %d messages, not 1 or 2",
+						backlogLastId, "+", key, len(msgs))
+				}
+
+				msg := msgs[0]
+				entity, err := makeEntity(msg.Values)
+				if err != nil {
+					return errors.Wrapf(err, "can't structify backlog value %q for %q", backlogLastId, key)
+				}
+
+				if callback(entity) {
+					out <- msg
+					backlogMsgCounter++
+					telemetry.Stats.Callback.Add(1)
+
+					if len(msgs) == 1 {
+						backlogLastId = ""
+						logger.Infow("Finished rolling back backlog of callback elements", zap.Int("delay", backlogMsgCounter))
+					} else {
+						backlogLastId = msgs[1].ID
+						backlogTimerInterval = backlogTimerMinInterval
+						_ = backlogTimer.Reset(backlogTimerInterval)
```

Comment:
Wont this effectively do a `sleep(10ms)`? So if everything works fine, won't this artificially limit throughput?